src/QsCompiler/Transformations/CallGraphWalker.cs

```diff
@@ -17,6 +16,258 @@ namespace Microsoft.Quantum.QsCompiler.Transformations.CallGraphWalker
     using ExpressionKind = QsExpressionKind<TypedExpression, Identifier, ResolvedType>;
     using ResolvedTypeKind = QsTypeKind<ResolvedType, UserDefinedType, QsTypeParameter, CallableInformation>;
 
+    /// <summary>
+    /// Class for using Johnson's algorithm for finding all simple cycles in a graph.
+    /// A simple cycle is one where no node repeats in the cycle until getting back
+    /// to the beginning of the cycle.
+    ///
+    /// The algorithm uses Tarjan's algorithm for finding all strongly-connected
+    /// components in a graph. A strongly-connected component, or SCC, is a subgraph
+    /// in which all nodes can reach all other nodes in the subgraph.
+    ///
+    /// The algorithm begins by using Tarjan's algorithm to find the SCCs of the graph.
+    /// Each SCC is processed for cycles, using the minimum-valued node of the SCC as the
+    /// starting node. Then the starting node is removed from the graph, and Tarjan is used
+    /// again to find SCCs of the new graph, and the next SCC is processed. This repeats
+    /// until there are no more SCCs (and no more nodes) left in the graph.
+    ///
+    /// The starting point for this algorithm in this class is the GetAllCycles method.
+    /// </summary>
+    internal class JohnsonCycleFind
+    {
+        private Stack<(HashSet<int> SCC, int MinNode)> SccStack = new Stack<(HashSet<int> SCC, int MinNode)>();
+
+        /// <summary>
+        /// Johnson's algorithm for finding all cycles in a graph.
+        /// This uses Tarjan's algorithm for finding strongly-connected components, or SCCs,
+        /// of the graph and processes those SCCs individually for finding cycles.
+        ///
+        /// This returns a list of cycles, each represented as a list of nodes. Nodes
+        /// for this algorithm are integers. The cycles are guaranteed to not contain
+        /// any duplicates and the last node in the list is assumed to be connected
+        /// to the first.
+        /// </summary>
+        public List<List<int>> GetAllCycles(Dictionary<int, List<int>> graph)
```

Comment:
The purpose is to find an return all the cycles in the call graph. Cycles are generally okay, as this would indicate recursion, which is allowed. However we need to analyze the cycles for certain qualities involving the type arguments used to in the call cycle. This code for this analysis is being developed separately by @bettinaheim .

Reply:
I'm just curious, could you give an example of a problematic cycle? Maybe, link to the issue# that has lead this work?

Reply:
Also, if most cycles aren't going to be problematic, enumerating all of them would be wasteful.

Reply:
I'm not sure the issue is written down anywhere in any detail. If you want to know more, the best way would probably be to have a quick call with me or Bettina or both.

Reply:
Resolved offline: we want to go for a general solution that can be used at a later time as a template for a more specific solution that may be able to take advantage of scenario-specific optimizations.